Refactor processing workflow tests

jraddaoui opened the pull request:
Split large test file into three and introduce a reusable expectation system to reduce code duplication, improve reusability, and enable easier extension. The expectation functions provide parameterized mock setups for common workflow activities.

- Extract expectation functions for reusable activity mocks.
- Centralize test constants and variables in expectations file.
- Separate test setup logic into dedicated setup file.
- Add ExecuteAndValidateWorkflow helper.

jraddaoui committed:
Refactor processing workflow tests

Split large test file into three and introduce a reusable expectation
system to reduce code duplication, improve reusability, and enable
easier extension. The expectation functions provide parameterized
mock setups for common workflow activities.

- Extract expectation functions for reusable activity mocks.
- Centralize test constants and variables in expectations file.
- Separate test setup logic into dedicated setup file.
- Add ExecuteAndValidateWorkflow helper.

jraddaoui commented:
Wait for it, I thought it would be better to ask you from some feedback at this point, but there are only a few remaining things to do, so I'll finish and explain it all at once. Sorry for the noise.

jraddaoui commented:
Ready for review when you have some time @djjuhasz. I think it makes no sense looking at the diff, I'd check the three test files individually. In the setup file I have only moved the existing functions and added a helper to execute and validate the workflow. In the other two is where most of the changes are happening, preprocessing_test now only contains the test cases, and preprocessing_expectations_test includes the new expectations and variables.

I've moved all the variables we had duplicated on each test alongside the other global vars/const and added expectation functions for those that were used more than once across all tests. I've used Copilot with Claude Sonnet 4 in agent mode in a few point in this changes, it was really helpful identifying those duplicated vars/expectations and adding comments but not so helpful with the parameterization and other aspects.

This was the length of the current test cases:

- TestConfirmation: 244 lines
- TestAutoApprovedAIP: 253 lines
- TestAMWorkflow: 249 lines
- TestRejection: 176 lines
- TestChildWorkflows: 343 lines
- TestFailedSIP: 178 lines
- TestFailedPIPA3m: 214 lines
- TestFailedPIPAM: 168 lines
- TestInternalUpload: 146 lines

And a summary:

- Shortest test: TestInternalUpload (146 lines)
- Longest test: TestChildWorkflows (343 lines)
- Average length: ~219 lines
- Total test lines: ~1,971 lines (excluding helper functions and imports)

Compared with these changes:

- TestConfirmation: 73 lines
- TestRejection: 67 lines
- TestAutoApprovedAIP: 58 lines
- TestAMWorkflow: 88 lines
- TestChildWorkflows: 141 lines
- TestFailedSIP: 56 lines
- TestFailedPIPA3m: 61 lines
- TestFailedPIPAM: 45 lines
- TestInternalUpload: 68 lines

Summary

- Shortest test: TestFailedPIPAM (45 lines)
- Longest test: TestChildWorkflows (141 lines)
- Average length: ~73 lines
- Total test lines: 657 lines (excluding helper functions and imports)

Hopefully this makes adding new tests a little less overwhelming.

jraddaoui committed:
Re-organize processing workflow

Move review process to the a3m branch, create updateSIPProcessing
function to avoid duplication and fix comments.

djjuhasz commented on internal/workflow/processing_expectations_test.go:
```diff
@@ -0,0 +1,501 @@
+// Each key represents a logical activity name that corresponds to a Temporal activity.
+// Functions use global constants and variables for consistent test data, with customization
+// through expectationParams.
+var expectations = map[string]expectationFunc{
```
I think it would be better to create a copy of the `expectations` map in each test to allow mutating the expectations for a specific test without changing the outcome of other test and to avoid contention problems when mutating the map.

djjuhasz replied:
Huh, it looks like testify doesn't support running tests in parallel (https://github.com/stretchr/testify?tab=readme-ov-file#suite-package), so I guess there's no risk of a simultaneous read and write to the map. :thinking:

jraddaoui replied:
But why would you want to mutate those expectations in test? The idea was to use them as they are with some minor modification through parameters, if you need a different expectation you can use it as before in the tests. For example for the download, I added the most common download activity call, other downloads just use their own expectation in test:

https://github.com/artefactual-sdps/enduro/blob/dev/processing-tests/internal/workflow/processing_expectations_test.go#L259-L277
https://github.com/artefactual-sdps/enduro/blob/dev/processing-tests/internal/workflow/processing_test.go#L609-L616

djjuhasz replied:
I was thinking a dev would modify the expectations to change the expected outcome of an activity - to make an activity fail for instance.  You're right that you can just write out the full activity mock like we did previously to change the expectations but it was not clear to me from the expectation structure or comments that it's meant to be a read-only data structure. There's also no protection in place to discourage a dev from mutating the map.

jraddaoui commented:
Added an extra commit to include a couple of expectation wrappers for a3m and added comments to the tests to specify what's being tested. A couple of new tests added using this expectations can be seen here: https://github.com/artefactual-sdps/enduro/pull/1376/commits/0c6990c54009ce75a073882520169af4c6a83087.

jraddaoui commented:
With the latest changes, including the tests mentioned above:

- TestConfirmation: 59 lines
- TestRejection: 52 lines
- TestAutoApprovedAIP: 29 lines
- TestAMWorkflow: 96 lines
- TestChildWorkflows: 126 lines
- TestFailedSIP: 63 lines
- TestFailedPIPA3m: 77 lines
- TestFailedPIPAM: 50 lines
- TestInternalUpload: 65 lines
- TestInternalUploadError: 74 lines
- TestSIPSourceUpload: 59 lines

Summary:

- Shortest test: TestAutoApprovedAIP (29 lines)
- Longest test: TestChildWorkflows (126 lines)
- Average length: ~68 lines
- Total test lines: ~750 lines (excluding helper functions and imports)

djjuhasz commented on internal/workflow/processing_expectations_test.go:
```diff
@@ -0,0 +1,501 @@
+	// Task identifier for task operations.
+	// Used in "createTask" and "completeTask" expectations.
+	// Required.
+	taskID int
```
Having a parameter that is required but only for a couple of expectations seems like it could lead to errors and be tricky to maintain (e.g. changing which expectations use the param will require updates in `preprocessing_test.go` and the comment here).  I think it would be better to use a constructor of some sort that sets a default value for `taskID` and other params where the "zero" value doesn't make sense as a default.

jraddaoui replied:
This particular field could even be removed and use always 0 as the task ID because it's used as return param and then as the input, but this is closer to reality where all tasks would get a different ID. But I'm not sure I understand your suggestion, each call to an expectation uses a new params instance, where would you add a constructor?

> (e.g. changing which expectations use the param will require updates in preprocessing_test.go and the comment here)

I think this will happen more than once and I don't see an issue with it. We'll probably add more expectations and parameters at some point.

djjuhasz replied:
Because the `expectationParams` is a single struct I would expect to instantiate it once with all of the parameters needed for a test, and then reuse it for each activity.  If the parameters are meant to be specific to an activity, then I would expect each activity to have it's own parameters struct.

jraddaoui replied:
What would you prefer?

djjuhasz replied:
I'd prefer it being a single struct (as now) but having a `defaultParams()` constructor that returns a new instance with defaults for all the fields.

jraddaoui replied:
Lets give it a try ;)

jraddaoui replied:
Thanks @djjuhasz! I like the idea, let me know how it looks to you.

djjuhasz reviewed (changes requested):
Very nice @jraddaoui! I made a couple of notes, but my main request is to make the expectations map a local variable instead of a global.

jraddaoui commented:
Thanks @djjuhasz!

djjuhasz reviewed (approved):
LGTM :+1:
